deducers.go

```diff
@@ -0,0 +1,292 @@
+	return cm.ctx
+}
+
+func (cm *callManager) run(ci callInfo) (context.Context, error) {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	if cm.ctx.Err() != nil {
+		// We've already been canceled; error out.
+		return nil, cm.ctx.Err()
+	}
+
+	if existingInfo, has := cm.running[ci]; has {
+		existingInfo.count++
+		cm.running[ci] = existingInfo
+	} else {
+		cm.running[ci] = timeCount{
+			count: 1,
+			start: time.Now(),
+		}
+	}
+
+	return cm.ctx, nil
+}
+
+func (cm *callManager) done(ci callInfo) {
+	cm.mu.Lock()
+
+	existingInfo, has := cm.running[ci]
+	if !has {
+		panic(fmt.Sprintf("sourceMgr: tried to complete a call that had not registered via run()"))
+	}
+
+	if existingInfo.count > 1 {
+		// If more than one is pending, don't stop the clock yet.
+		existingInfo.count--
+		cm.running[ci] = existingInfo
+	} else {
+		// Last one for this particular key; update metrics with info.
+		durCnt := cm.ran[ci.typ]
+		durCnt.count++
+		durCnt.dur += time.Now().Sub(existingInfo.start)
+		cm.ran[ci.typ] = durCnt
+		delete(cm.running, ci)
+	}
+
+	cm.mu.Unlock()
+}
+
+type callType uint
+
+const (
+	ctHTTPMetadata callType = iota
+	ctListVersions
+	ctGetManifestAndLock
+)
+
+// callInfo provides metadata about an ongoing call.
+type callInfo struct {
+	name string
+	typ  callType
+}
+
+type srcReturnChans struct {
+	ret chan *sourceGateway
+	err chan error
+}
+
+func (retchans srcReturnChans) awaitReturn() (*sourceGateway, error) {
+	select {
+	case sa := <-retchans.ret:
+		return sa, nil
+	case err := <-retchans.err:
+		return nil, err
+	}
+}
+
+type sourceCoordinator struct {
+	callMgr   *callManager
+	srcmut    sync.RWMutex // guards srcs and nameToURL maps
+	srcs      map[string]*sourceGateway
+	nameToURL map[string]string
+	psrcmut   sync.Mutex // guards protoSrcs map
+	protoSrcs map[string][]srcReturnChans
+	deducer   *deductionCoordinator
+	cachedir  string
+}
+
+func newSourceCoordinator(cm *callManager, deducer *deductionCoordinator, cachedir string) *sourceCoordinator {
+	return &sourceCoordinator{
+		callMgr:   cm,
+		deducer:   deducer,
+		cachedir:  cachedir,
+		srcs:      make(map[string]*sourceGateway),
+		nameToURL: make(map[string]string),
+		protoSrcs: make(map[string][]srcReturnChans),
+	}
+}
+
+func (sc *sourceCoordinator) getSourceGatewayFor(ctx context.Context, id ProjectIdentifier) (*sourceGateway, error) {
+	normalizedName := id.normalizedSource()
+
+	sc.srcmut.RLock()
+	if url, has := sc.nameToURL[normalizedName]; has {
+		if srcGate, has := sc.srcs[url]; has {
+			sc.srcmut.RUnlock()
+			return srcGate, nil
+		}
+	}
+	sc.srcmut.RUnlock()
+
+	// No gateway exists for this path yet; set up a proto, being careful to fold
+	// together simultaneous attempts on the same path.
+	rc := srcReturnChans{
+		ret: make(chan *sourceGateway),
+		err: make(chan error),
+	}
+
+	// The rest of the work needs its own goroutine, the results of which will
+	// be re-joined to this call via the return chans.
+	go func() {
```

Comment:
Thanks for the review!

Yeah, I've been leaning in that direction; I went heavy pseudo-actor model initially, and am slowly walking it back. The general issue I have with it is distinguishing between things that _should_ be called by the wrapper object, vs those that should not. That's a smaller design consideration, though - like I said, I'm walking it back :)

This one isn't shouldn't be a problem to make its own method for, but the other one (in `deductionCoordinator.deduce()` is absolutely not threadsafe unless it's being serialized through that channel.